pkg/services/ngalert/notifier/receiver_svc.go

```diff
@@ -271,26 +279,162 @@ func (rs *ReceiverService) DeleteReceiver(ctx context.Context, uid string, orgID
+	if err := rs.authz.AuthorizeCreate(ctx, user, r); err != nil {
+		return nil, err
+	}
+
+	revision, err := rs.cfgStore.Get(ctx, orgID)
+	if err != nil {
+		return nil, err
+	}
+
+	createdReceiver := r.Clone()
+	err = createdReceiver.Encrypt(rs.encryptor(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	err = revision.CreateReceiver(&createdReceiver)
+	if err != nil {
+		return nil, err
+	}
+	createdReceiver.Version = createdReceiver.Fingerprint()
+
+	err = rs.xact.InTransaction(ctx, func(ctx context.Context) error {
+		err = rs.cfgStore.Save(ctx, revision, orgID)
+		if err != nil {
+			return err
+		}
+		return rs.setReceiverProvenance(ctx, orgID, &createdReceiver)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &createdReceiver, nil
+}
+
+func (rs *ReceiverService) UpdateReceiver(ctx context.Context, r *models.Receiver, storedSecureFields map[string][]string, orgID int64, user identity.Requester) (*models.Receiver, error) {
+	if err := rs.authz.AuthorizeUpdate(ctx, user, r); err != nil {
+		return nil, err
+	}
+
+	revision, err := rs.cfgStore.Get(ctx, orgID)
+	if err != nil {
+		return nil, err
+	}
+	postable, err := revision.GetReceiver(r.GetUID())
+	if err != nil {
+		return nil, err
+	}
+
+	storedProvenances, err := rs.provisioningStore.GetProvenances(ctx, orgID, (&definitions.EmbeddedContactPoint{}).ResourceType())
+	if err != nil {
+		return nil, err
+	}
+	existing, err := PostableApiReceiverToReceiver(postable, getReceiverProvenance(storedProvenances, postable))
+	if err != nil {
+		return nil, err
+	}
+
+	// Check optimistic concurrency.
+	err = rs.checkOptimisticConcurrency(existing, r.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := rs.validator(existing.Provenance, r.Provenance); err != nil {
+		return nil, err
+	}
+
+	// We need to perform two important steps to process settings on an updated integration:
+	// 1. Encrypt new or updated secret fields as they will arrive in plain text.
+	// 2. For updates, callers do not re-send unchanged secure settings and instead mark them in SecureFields. We need
+	//      to load these secure settings from the existing integration.
+	updatedReceiver := r.Clone()
+	err = updatedReceiver.Encrypt(rs.encryptor(ctx))
+	if err != nil {
+		return nil, err
+	}
+	if len(storedSecureFields) > 0 {
+		updatedReceiver.WithExistingSecureFields(existing, storedSecureFields)
+	}
+
+	err = revision.UpdateReceiver(&updatedReceiver)
+	if err != nil {
+		return nil, err
+	}
+	updatedReceiver.Version = updatedReceiver.Fingerprint()
+
+	err = rs.xact.InTransaction(ctx, func(ctx context.Context) error {
+		// If the name of the receiver changed, we must update references to it in both routes and notification settings.
+		if existing.Name != r.Name {
+			affected, err := rs.ruleNotificationsStore.RenameReceiverInNotificationSettings(ctx, orgID, existing.Name, r.Name)
```

Comment:
I think this should be a little bit more complex. we should check provenance status compatibility. For example, if we rename a receiver via UI but rules are provisioned, this call should be rejected.